### PR TITLE
ENH: Added directed_hausdorff from scipy to speed up hausdorff calc

### PR DESCRIPTION
**Pull request by tylerjereddy**
This PR attempts to use [`scipy.spatial.distance.directed_hausdorff`](https://docs.scipy.org/doc/scipy/reference/generated/generated/scipy.spatial.distance.directed_hausdorff.html) under the hood to speed up the MDA `hausdorff` function. However, even though the docstring of the latter suggests that the scipy implementation may be used directly on P & Q, this is not the case because of the (again, somewhat confusing) support for both 2D and 3D numpy array shapes in MDA psa module.

I made various attempts at reshaping P & Q prior to calling the scipy version, but none of them allowed unit tests to pass. I suggest that someone provides a clear explanation (with working code) of how I can filter the input arrays to scipy to allow the unit tests to pass. They *have* to be 2D numpy arrays and reshaping with i.e., (-1,3) doesn't work.

Or, we could abandon the idea of calling the scipy version, but then clean up the docstring accordingly, because a simple substitution of P & Q as described does not work at the moment in the context of the unit tests / shape requirements.

### CODE REVIEW

**Comment by tylerjereddy** [comment]
The best approach might be for someone to just write a helper function upstream that processes P & Q & returns them (or views of them) that are always in the appropriate shape / configuration for my purposes here. Then I can rebase on that and continue working here.

**Comment by orbeckst** [comment]
https://www.mdanalysis.org/benchmarks/ now also has the historical PSA benchmark data included.

**Comment by orbeckst** [comment]
@tylerjereddy I don't think that the doc string for [psa.hausdorff()](https://www.mdanalysis.org/mdanalysis/documentation_pages/analysis/psa.html#MDAnalysis.analysis.psa.hausdorff) says that the scipy function can be used directly – even though this would have been nice.

@sseyler should correct me here – but I think all that is needed is too use what the PSA docs call the "flattened" (2D) form where you consider points in the 3N dimensional configuration space of the N atoms:
```python
P_flat = P.reshape(len(P), -1)
```
(or use `P.shape[0]` instead of `len(P)` if you like it more explicit.)

For example for a path with 20 vertices of a protein with 100 atoms:
```python
import numpy as np
Np = 20
N = 100
P = np.random.random((Np, N, 3))   # "full" path in PSA, 3d array
print(P.shape)
# (20, 100, 3)

P_flat = P.reshape(len(P), -1)       # "flattened" path in PSA, 2d array
print(P_flat.shape)
# (20, 300)
```

(And you can get the 3d back with `P_flat.reshape(len(P), -1, 3)`.)

**Comment by tylerjereddy** [comment]
The docs do call directed_hausdorff directly on variables with the same name as the input arguments to the function, though that may not have been the conveyed intention. P, Q probably shouldn't be used in that scenario then.

It may be a straigthforward reshaping in the end, but I'll probably defer it to @sseyler or someone else to confirm / code it in upstream because every time I read those shape explanations I get confused!

**Comment by tylerjereddy** [comment]
@orbeckst I tried two different permutations of your proposed reshaping without success getting unit tests to pass.

**Comment by tylerjereddy** [comment]
The first of the commits might be bad because of the repeated reshape, but the second I'm not clear on what that wouldn't do what is being intended.

**Comment by orbeckst** [comment]
Maybe the difference is that your scipy version uses a Euclidean distance and PSA uses RMSD, i.e. Euclidean/sqrt(N)?

**Comment by tylerjereddy** [comment]
@orbeckst Yes, that was the problem. The function docstring should probably clarify the distance metric being used is RMSD, regardless of what other docs in this module say. This is compounded by a direct comparison with the scipy implementation that uses the Euclidean metric in Notes section of the function docstring.

My current fix is pretty ugly -- it might be possible to reuse the old axis mangling code now though. Conversely, the tests do pass now -- this probably just means that the exception handling isn't currently unit tested.

I think this mean that `N` has to be the same for both `P` & `Q`?

**Comment by orbeckst** [comment]
`N` has to be the same for P and Q because `3N` is the dimension of the space in which the two curves live.

I agree that the docs should explicitly state that this function uses RMSD as point-wise metric. This is particular important because otherwise people might wonder how they can easily slot in another point-wise metric – something that is an interesting idea for PSA but we haven't explored and clearly the PSA module would need some enhancements to support custom point-wise metrics.

**Comment by tylerjereddy** [comment]
Ok, I'll try to update the docs while I work on this PR then. Probably just one or two extra words will suffice. If we manage to switch to scipy backend (it now looks pretty convincing that we can), the docstring can be shortened a fair bit anyway without needing to make the scipy comparison.

I should add a test that an exception is shown for an array with > 3 dims (a, b, c, d) just to be thorough as well so that crude patching code like I've done here isn't sufficient in the future, etc.

I think I just got frustrated because I was hoping for some low-hanging fruit for an enhacement! But, this won't be so bad in the end, I don't think.

**Comment by tylerjereddy** [comment]
I had one or two more adjustments in mind + would be good to actually do the `asv` benchmark (just for two commits) as a sample of how "perf improvement" PRs might go.

**Comment by sseyler** [comment]
@tylerjereddy IIRC, I wrote the functions to accept these "flattened" arrays so I could use them with my Cython code; my Cython code used the memoryview notation and I had some kind of trouble/performance hit when using 3D vs. 2D arrays for the inputted paths. Maybe it's best to deal with paths in a more general way:

Given a path `P` with `T` steps (frames) and `N` dimensions, the input array would be of shape `(T,N)`. Then, for another path `Q` using the convention `T` steps, `n` atoms, and `3` (Euclidean) coordinates per atom, with shape `(T,n,3)`, the path should be flattened to get the shape `(T,3*n)`. Perhaps this 2D array should be the default shape for input paths so that paths with the `(T,n,3)` convention should be converted from a 3D array to a 2D array (rather than the other way around)? I believe this is how you're doing it currntly with the `hausdorff` function, though the `hausdorff_wavg` functions and the like use my old convention (in the `get_coord_axes` helper function).

**Comment by tylerjereddy** [comment]
I've made some of the adjustments I wanted to (including docstring clarifications) -- please have a look. Here's the performance summary vs. a previous commit in develop using Python 2.7:

`asv continuous --bench "\btime_hausdorff\b" ad3d811729 f08bdcfbf2d`


```
      before           after         ratio
     [ad3d8117]       [f08bdcfb]
+        96.1±2μs          120±2μs     1.25  analysis.psa.PSA_metricBench.time_hausdorff(10, 5)
+         108±2μs          125±4μs     1.16  analysis.psa.PSA_metricBench.time_hausdorff(10, 25)
-      1.23±0.1ms         516±20μs     0.42  analysis.psa.PSA_metricBench.time_hausdorff(100, 5)
-      2.38±0.1ms         607±10μs     0.26  analysis.psa.PSA_metricBench.time_hausdorff(100, 25)
-      3.70±0.2ms         938±30μs     0.25  analysis.psa.PSA_metricBench.time_hausdorff(200, 5)
-      3.84±0.1ms         866±20μs     0.23  analysis.psa.PSA_metricBench.time_hausdorff(100, 50)
-      8.48±0.3ms      1.42±0.05ms     0.17  analysis.psa.PSA_metricBench.time_hausdorff(200, 25)
-      14.6±0.5ms      2.45±0.08ms     0.17  analysis.psa.PSA_metricBench.time_hausdorff(200, 50)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

Note that for really small inputs there's actually a modest slow-down, perhaps because of the necessary up-front data processing to accommodate the data shape requirements for scipy. Some reasonable performance improvements for larger data sizes though.

**Comment by orbeckst** [comment]
@tylerjereddy if you're happy with the PR, remove the WIP name/tag and ping me. Or you can merge yourself.

**Comment by tylerjereddy** [comment]
Ok, squash merging & yes, I suspect there are other opportunities for a speed up with the pair calculation. I'm guessing the weighted variants would be harder to leverage the scipy stuff directly though.

File: package/MDAnalysis/analysis/psa.py [hunk 1]
```diff
@@ -218,6 +218,7 @@
 
 import numpy as np
 from scipy import spatial, cluster
+from scipy.spatial.distance import directed_hausdorff
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/psa.py]: what is the minimum scipy version we depend on now?

Comment (tylerjereddy) [reply package/MDAnalysis/analysis/psa.py]: Shouldn't be hard to check -- maybe making it the "stable" 1.0 version would be sensible soon.

Comment (kain88-de) [reply package/MDAnalysis/analysis/psa.py]: This function entered scipy in 0.19.0 according to the docs. So I think we should raise our minimum required version to this one.

Comment (orbeckst) [reply package/MDAnalysis/analysis/psa.py]: @kain88-de thanks for catching this!

**Pull request merged**